node.cpp

```diff
@@ -0,0 +1,49 @@
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+#include "node.h"
+
+#define MAX_CHILDREN 100
+
+Node *node_new(const char *name)
+{
+  Node *node = (Node *)calloc(1, sizeof(Node));
+  node->internal_type = name;
+  return node;
+}
+
+void node_add_role(Node *node, uint16_t)
+{
+}
+
+void node_add_child(Node *parent, Node *child)
+{
+  if (!parent || !child)
+  {
+    return;
+  }
+  if (!parent->children)
+  {
+    parent->children = (Node **)malloc(MAX_CHILDREN * sizeof(Node *));
```

Comment:
Using a vector would make most of this code unneeded and would just be a line `parent->children.push(child)`